src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -187,53 +193,95 @@ private void doStart() throws Exception {
+    private static class MaskingDecorator extends TaskListenerDecorator {
+        @Nonnull
+        static TaskListenerDecorator createFrom(@Nonnull Pattern pattern, @Nonnull String charsetName, @CheckForNull TaskListenerDecorator original) {
+            return TaskListenerDecorator.merge(new MaskingDecorator(new SecretPattern(pattern), charsetName), original);
+        }
+
+        private static final long serialVersionUID = 1;
+
+        private final SecretPattern pattern;
+        private final String charsetName;
+
+        private MaskingDecorator(SecretPattern pattern, String charsetName) {
+            this.pattern = pattern;
+            this.charsetName = charsetName;
+        }
+
+        @Nonnull
+        @Override
+        public OutputStream decorate(@Nonnull OutputStream logger) throws IOException, InterruptedException {
             return new LineTransformationOutputStream() {
-                @Override protected void eol(byte[] b, int len) throws IOException {
-                    if (!p.toString().isEmpty()) {
-                        Matcher m = p.matcher(new String(b, 0, len, charsetName));
-                        if (m.find()) {
-                            logger.write(m.replaceAll("****").getBytes(charsetName));
-                        } else {
-                            // Avoid byte → char → byte conversion unless we are actually doing something.
-                            logger.write(b, 0, len);
-                        }
+                @Override
+                protected void eol(byte[] b, int len) throws IOException {
+                    Matcher m = pattern.matcher(new String(b, 0, len, charsetName));
+                    if (m.find()) {
+                        logger.write(m.replaceAll("****").getBytes(charsetName));
                     } else {
                         // Avoid byte → char → byte conversion unless we are actually doing something.
                         logger.write(b, 0, len);
                     }
                 }
-                @Override public void flush() throws IOException {
+
+                @Override
+                public void flush() throws IOException {
                     logger.flush();
                 }
-                @Override public void close() throws IOException {
+
+                @Override
+                public void close() throws IOException {
                     super.close();
                     logger.close();
                 }
             };
         }
+    }
+
+    /** @deprecated Use {@link MaskingDecorator} */
+    @Deprecated
+    private static final class Filter extends ConsoleLogFilter implements Serializable {
+
+        private static final long serialVersionUID = 1;
+
+        @Override public OutputStream decorateLogger(AbstractBuild _ignore, final OutputStream logger) throws IOException, InterruptedException {
+            throw new UnknownServiceException();
```

Comment:
??

Reply:
Typo most likely; I must've hit tab after typing `throw new Un`. Should be `UnsupportedOperationException`.

Reply:
I think you must retain compatibility here. If someone has a build running with credential masking active, upgrades the plugin, and restarts Jenkins, this method will be called after the restart.